File: backend/app/services/report_pdf.py
"""Monthly / quarterly income-expense report as an A4 PDF.

Layout follows the Giai đoạn 4 brief: header (farm, address, period, date),
summary (thu, chi, lãi/lỗ), income lines, expense lines, notes derived from
the data, footer. Text and numbers only, no images.

Open Sans (bundled for the phone in mobile/src/assets/fonts) is embedded so
Vietnamese diacritics and the đồng sign render everywhere — Helvetica, the
PDF core font, cannot show "Nguyễn" or "₫".

The same layout is produced on the phone from HTML
(mobile/src/domain/reportPdf.ts); `report_notes()` here and `reportNotes()`
there implement the identical rules so both documents say the same thing.
"""

from __future__ import annotations

from dataclasses import dataclass
from datetime import datetime, timedelta

from fpdf import FPDF

from app.core.config import REPO_ROOT
from app.services import ledger_service as ledger

FONT_DIR = REPO_ROOT / "mobile" / "src" / "assets" / "fonts"
FONT_REGULAR = FONT_DIR / "OpenSans-Regular.ttf"
FONT_BOLD = FONT_DIR / "OpenSans-Bold.ttf"

SUPPORT_EMAIL = "support@agrilog.vn"
APP_NAME = "AgriLog v2"


def vnd(value: float) -> str:
    return f"{int(round(value)):,}".replace(",", ".") + "₫"


def signed_vnd(value: float) -> str:
    return ("+" if value >= 0 else "−") + vnd(abs(value))


def day_label(occurred_at: int) -> str:
    return datetime.fromtimestamp(occurred_at / 1000, tz=ledger.VN_TZ).strftime("%d/%m")


@dataclass
class ReportNotes:
    fertilizer_lines: list[tuple[str, float]]
    fertilizer_total: float
    stock_kg: float
    stock_value: float
    stock_kinds: int
    avg_monthly_fertilizer: float | None
    months_sampled: int

    def as_lines(self) -> list[str]:
        lines: list[str] = []
        if self.fertilizer_lines:
            parts = " + ".join(vnd(v) for _, v in self.fertilizer_lines)
            lines.append(f"Chi phí phân bón: {parts} = {vnd(self.fertilizer_total)}")
        else:
            lines.append("Chi phí phân bón: không có khoản chi phân bón trong kỳ")
        if self.stock_kinds > 0:
            lines.append(
                f"Còn dư phân bón: {self.stock_kg:g} kg ({self.stock_kinds} loại, ≈ {vnd(self.stock_value)} theo giá nhập bình quân)"
            )
        else:
            lines.append("Còn dư phân bón: kho trống")
        if self.avg_monthly_fertilizer is not None and self.avg_monthly_fertilizer > 0:
            lines.append(
                f"Khuyến nghị: chi phân bón trung bình {vnd(self.avg_monthly_fertilizer)}/tháng "
                f"(tính trên {self.months_sampled} tháng gần nhất) — nên giữ dự phòng khoảng "
                f"{vnd(self.avg_monthly_fertilizer)} mỗi tháng"
            )
        else:
            lines.append("Khuyến nghị: chưa đủ lịch sử chi phân bón để gợi ý mức dự phòng")
        return lines


def report_notes(
    report: ledger.FinancialReport,
    stock_lines: list[ledger.StockLine],
    all_expenses: list,
    period: ledger.Period,
) -> ReportNotes:
    """Notes are computed from the ledger, never typed in: the fertiliser
    costs of the period, what is left in the shed, and a buffer suggestion
    equal to the average fertiliser spend over the three months ending with
    the period."""
    fert = [(r.description, r.amount) for r in report.expenses if r.kind == "fertilizer"]
    in_stock = [l for l in stock_lines if l.stock_kg > 0]

    # Three calendar months ending with the report's last month.
    last_month = period.month if period.month is not None else (period.quarter or 1) * 3
    window_end = datetime(period.year, last_month, 1, tzinfo=ledger.VN_TZ)
    window_end = (window_end + timedelta(days=32)).replace(day=1)
    window_start = window_end
    for _ in range(3):
        window_start = (window_start - timedelta(days=1)).replace(day=1)
    lo, hi = int(window_start.timestamp() * 1000), int(window_end.timestamp() * 1000)

    window_total = sum(r.amount for r in all_expenses if r.kind == "fertilizer" and lo <= r.occurred_at < hi)
    months_with_data = len(
        {
            datetime.fromtimestamp(r.occurred_at / 1000, tz=ledger.VN_TZ).strftime("%Y-%m")
            for r in all_expenses
            if r.kind == "fertilizer" and lo <= r.occurred_at < hi
        }
    )

    return ReportNotes(
        fertilizer_lines=fert,
        fertilizer_total=sum(v for _, v in fert),
        stock_kg=sum(l.stock_kg for l in in_stock),
        stock_value=sum(l.stock_value for l in in_stock),
        stock_kinds=len(in_stock),
        avg_monthly_fertilizer=(window_total / 3) if months_with_data else None,
        months_sampled=3,
    )


class _Report(FPDF):
    def __init__(self) -> None:
        super().__init__(orientation="P", unit="mm", format="A4")
        self.set_margins(20, 20, 20)
        self.set_auto_page_break(auto=True, margin=20)
        self.add_font("OpenSans", "", str(FONT_REGULAR))
        self.add_font("OpenSans", "B", str(FONT_BOLD))

    def rule(self) -> None:
        self.set_draw_color(229, 231, 235)
        self.set_line_width(0.3)
        y = self.get_y()
        self.line(20, y, 190, y)
        self.ln(3)

    def heading(self, text: str) -> None:
        self.ln(3)
        self.set_font("OpenSans", "B", 12)
        self.set_text_color(17, 24, 39)
        self.cell(0, 8, text, new_x="LMARGIN", new_y="NEXT")
        self.rule()

    def row(self, left: str, right: str, bold: bool = False, muted: bool = False) -> None:
        self.set_font("OpenSans", "B" if bold else "", 10)
        self.set_text_color(107, 114, 128) if muted else self.set_text_color(17, 24, 39)
        self.cell(120, 6.5, left)
        self.cell(50, 6.5, right, align="R", new_x="LMARGIN", new_y="NEXT")

    def line_item(self, day: str, description: str, amount: float) -> None:
        self.set_font("OpenSans", "", 10)
        self.set_text_color(17, 24, 39)
        self.cell(16, 6.5, day)
        self.cell(104, 6.5, _clip(description, 60))
        self.cell(50, 6.5, vnd(amount), align="R", new_x="LMARGIN", new_y="NEXT")


def _clip(value: str, width: int) -> str:
    return value if len(value) <= width else value[: width - 1] + "…"


def build_pdf(
    *,
    farmer_name: str,
    address: str | None,
    report: ledger.FinancialReport,
    notes: ReportNotes,
    generated_at: datetime,
) -> bytes:
    pdf = _Report()
    pdf.add_page()

    pdf.set_font("OpenSans", "B", 16)
    pdf.set_text_color(46, 111, 64)
    pdf.cell(0, 10, "BÁO CÁO THU – CHI NÔNG HỘ", align="C", new_x="LMARGIN", new_y="NEXT")
    pdf.set_font("OpenSans", "", 10)
    pdf.set_text_color(17, 24, 39)
    pdf.cell(0, 6, f"Nông hộ: {farmer_name}", new_x="LMARGIN", new_y="NEXT")
    pdf.cell(0, 6, f"Địa chỉ: {address or 'Chưa cập nhật'}", new_x="LMARGIN", new_y="NEXT")
    pdf.cell(
        0,
        6,
        f"Kỳ: {report.period}  |  Ngày xuất: {generated_at.strftime('%d/%m/%Y')}",
        new_x="LMARGIN",
        new_y="NEXT",
    )
    pdf.rule()

    pdf.heading("TÓM TẮT")
    pdf.row("Thu", vnd(report.total_income))
    pdf.row("Chi", vnd(report.total_expense))
    verdict = "LÃI" if report.profit >= 0 else "LỖ"
    pdf.row(f"Lãi/Lỗ  [{verdict}]", signed_vnd(report.profit), bold=True)

    pdf.heading("CHI TIẾT THU")
    if report.incomes:
        for row in report.incomes:
            pdf.line_item(day_label(row.occurred_at), row.description, row.amount)
        pdf.row("Tổng thu", vnd(report.total_income), bold=True)
    else:
        pdf.row("Không có khoản thu trong kỳ", "", muted=True)

    pdf.heading("CHI TIẾT CHI")
    if report.expenses:
        for row in report.expenses:
            pdf.line_item(day_label(row.occurred_at), row.description, row.amount)
        pdf.row("Tổng chi", vnd(report.total_expense), bold=True)
    else:
        pdf.row("Không có khoản chi trong kỳ", "", muted=True)

    pdf.heading("GHI CHÚ")
    pdf.set_font("OpenSans", "", 10)
    pdf.set_text_color(17, 24, 39)
    for line in notes.as_lines():
        pdf.multi_cell(0, 6, f"• {line}", new_x="LMARGIN", new_y="NEXT")

    pdf.ln(6)
    pdf.rule()
    pdf.set_font("OpenSans", "", 9)
    pdf.set_text_color(107, 114, 128)
    pdf.cell(0, 6, f"Được tạo bởi: {APP_NAME}  |  Hỗ trợ: {SUPPORT_EMAIL}", align="C", new_x="LMARGIN", new_y="NEXT")

    return bytes(pdf.output())

File: mobile/src/domain/reportPdf.ts
/**
 * The monthly / quarterly report as self-contained HTML, rendered to an A4
 * PDF on the device by react-native-html-to-pdf. Same sections and the same
 * note rules as backend/app/services/report_pdf.py, so a farmer's PDF and
 * the one web-admin downloads say the same thing.
 *
 * Fonts: the WebView that prints the HTML uses the system sans font, which
 * on Android and iOS covers Vietnamese and "₫"; Helvetica is named only as
 * the fallback the brief asks for.
 */

import type {FinancialReport, LedgerEntry} from './finance';
import type {PeriodFilter, StockLine} from './warehouse';
import {periodBounds} from './warehouse';

export const SUPPORT_EMAIL = 'support@agrilog.vn';
export const APP_NAME = 'AgriLog v2';

export interface ReportNotes {
  fertilizerLines: {description: string; amount: number}[];
  fertilizerTotal: number;
  stockKg: number;
  stockValue: number;
  stockKinds: number;
  /** Average fertiliser spend per month over the 3 months ending with the period; null without history. */
  avgMonthlyFertilizer: number | null;
}

/** Three calendar months ending with the report's last month, as [start, end) ms. */
export function trailingWindow(period: PeriodFilter): [number, number] | null {
  if (period.kind === 'all') return null;
  const lastMonth = period.kind === 'month' ? period.month : period.quarter * 3;
  const end = new Date(period.year, lastMonth, 1).getTime();
  const start = new Date(period.year, lastMonth - 3, 1).getTime();
  return [start, end];
}

export function reportNotes(
  report: FinancialReport,
  stock: readonly StockLine[],
  allExpenses: readonly LedgerEntry[],
): ReportNotes {
  const fertilizerLines = report.expenses
    .filter(e => e.kind === 'fertilizer')
    .map(e => ({description: e.description, amount: e.amount}));
  const inStock = stock.filter(l => l.stockKg > 0);

  const window = trailingWindow(report.period);
  let avg: number | null = null;
  if (window) {
    const inWindow = allExpenses.filter(
      e => e.kind === 'fertilizer' && e.occurredAt >= window[0] && e.occurredAt < window[1],
    );
    if (inWindow.length > 0) avg = inWindow.reduce((s, e) => s + e.amount, 0) / 3;
  }

  return {
    fertilizerLines,
    fertilizerTotal: fertilizerLines.reduce((s, l) => s + l.amount, 0),
    stockKg: inStock.reduce((s, l) => s + l.stockKg, 0),
    stockValue: inStock.reduce((s, l) => s + l.stockValue, 0),
    stockKinds: inStock.length,
    avgMonthlyFertilizer: avg,
  };
}

export function noteLines(notes: ReportNotes, formatVnd: (v: number) => string): string[] {
  const lines: string[] = [];
  if (notes.fertilizerLines.length > 0) {
    lines.push(
      `Chi phí phân bón: ${notes.fertilizerLines.map(l => formatVnd(l.amount)).join(' + ')} = ${formatVnd(notes.fertilizerTotal)}`,
    );
  } else {
    lines.push('Chi phí phân bón: không có khoản chi phân bón trong kỳ');
  }
  if (notes.stockKinds > 0) {
    lines.push(
      `Còn dư phân bón: ${formatKg(notes.stockKg)} kg (${notes.stockKinds} loại, ≈ ${formatVnd(Math.round(notes.stockValue))} theo giá nhập bình quân)`,
    );
  } else {
    lines.push('Còn dư phân bón: kho trống');
  }
  if (notes.avgMonthlyFertilizer !== null && notes.avgMonthlyFertilizer > 0) {
    const avg = formatVnd(Math.round(notes.avgMonthlyFertilizer));
    lines.push(
      `Khuyến nghị: chi phân bón trung bình ${avg}/tháng (tính trên 3 tháng gần nhất) — nên giữ dự phòng khoảng ${avg} mỗi tháng`,
    );
  } else {
    lines.push('Khuyến nghị: chưa đủ lịch sử chi phân bón để gợi ý mức dự phòng');
  }
  return lines;
}

function formatKg(value: number): string {
  return new Intl.NumberFormat('vi-VN', {maximumFractionDigits: 2}).format(value);
}

export function signedVnd(value: number, formatVnd: (v: number) => string): string {
  return (value >= 0 ? '+' : '−') + formatVnd(Math.abs(value));
}

export function reportFileName(period: PeriodFilter): string {
  if (period.kind === 'month') return `bao-cao-thu-chi-${period.year}-thang-${period.month}`;
  if (period.kind === 'quarter') return `bao-cao-thu-chi-${period.year}-quy-${period.quarter}`;
  return 'bao-cao-thu-chi';
}

export interface ReportHtmlInput {
  farmerName: string;
  address: string | null;
  report: FinancialReport;
  notes: ReportNotes;
  generatedAt: number;
  formatVnd: (v: number) => string;
  formatDay: (at: number) => string;
}

const esc = (s: string) =>
  s.replace(/&/g, '&amp;').replace(/</g, '&lt;').replace(/>/g, '&gt;').replace(/"/g, '&quot;');

/** A4 report: header, summary, thu, chi, ghi chú, footer. Text and numbers only. */
export function reportHtml({farmerName, address, report, notes, generatedAt, formatVnd, formatDay}: ReportHtmlInput): string {
  const verdict = report.profit >= 0 ? 'LÃI' : 'LỖ';
  const rows = (entries: readonly LedgerEntry[]) =>
    entries
      .map(
        e =>
          `<tr><td class="day">${esc(formatDay(e.occurredAt).slice(0, 5))}</td><td>${esc(e.description)}</td><td class="num">${esc(formatVnd(e.amount))}</td></tr>`,
      )
      .join('');
  const empty = (label: string) => `<tr><td colspan="3" class="muted">${label}</td></tr>`;

  return `<!doctype html>
<html lang="vi"><head><meta charset="utf-8">
<style>
  @page { size: A4; margin: 20mm; }
  body { font-family: "Open Sans", Roboto, Helvetica, Arial, sans-serif; color: #111827; font-size: 11pt; margin: 0; }
  h1 { font-size: 16pt; color: #2E6F40; text-align: center; margin: 0 0 8pt; letter-spacing: .02em; }
  h2 { font-size: 12pt; margin: 16pt 0 4pt; padding-bottom: 3pt; border-bottom: 1px solid #E5E7EB; }
  .meta { margin: 0; line-height: 1.5; }
  hr { border: 0; border-top: 1px solid #E5E7EB; margin: 8pt 0; }
  table { width: 100%; border-collapse: collapse; }
  td { padding: 3pt 2pt; vertical-align: top; }
  td.day { width: 14mm; color: #6B7280; }
  td.num { text-align: right; white-space: nowrap; font-variant-numeric: tabular-nums; }
  tr.total td { font-weight: 700; border-top: 1px solid #E5E7EB; padding-top: 5pt; }
  .muted { color: #6B7280; }
  .summary td:first-child { width: 70%; }
  .verdict { font-weight: 700; }
  ul { margin: 4pt 0 0 14pt; padding: 0; }
  li { margin-bottom: 3pt; }
  footer { margin-top: 24pt; padding-top: 6pt; border-top: 1px solid #E5E7EB; color: #6B7280; font-size: 9pt; text-align: center; }
</style></head><body>
<h1>BÁO CÁO THU – CHI NÔNG HỘ</h1>
<p class="meta">Nông hộ: <strong>${esc(farmerName)}</strong><br>
Địa chỉ: ${esc(address ?? 'Chưa cập nhật')}<br>
Kỳ: ${esc(report.label)} &nbsp;|&nbsp; Ngày xuất: ${esc(formatDay(generatedAt))}</p>
<hr>
<h2>TÓM TẮT</h2>
<table class="summary">
<tr><td>Thu</td><td class="num">${esc(formatVnd(report.totalIncome))}</td></tr>
<tr><td>Chi</td><td class="num">${esc(formatVnd(report.totalExpense))}</td></tr>
<tr class="total"><td class="verdict">Lãi/Lỗ [${verdict}]</td><td class="num">${esc(signedVnd(report.profit, formatVnd))}</td></tr>
</table>
<h2>CHI TIẾT THU</h2>
<table>${report.incomes.length ? rows(report.incomes) + `<tr class="total"><td></td><td>Tổng thu</td><td class="num">${esc(formatVnd(report.totalIncome))}</td></tr>` : empty('Không có khoản thu trong kỳ')}</table>
<h2>CHI TIẾT CHI</h2>
<table>${report.expenses.length ? rows(report.expenses) + `<tr class="total"><td></td><td>Tổng chi</td><td class="num">${esc(formatVnd(report.totalExpense))}</td></tr>` : empty('Không có khoản chi trong kỳ')}</table>
<h2>GHI CHÚ</h2>
<ul>${noteLines(notes, formatVnd)
    .map(l => `<li>${esc(l)}</li>`)
    .join('')}</ul>
<footer>Được tạo bởi: ${APP_NAME} &nbsp;|&nbsp; Hỗ trợ: ${SUPPORT_EMAIL}</footer>
</body></html>`;
}

/** True when the period holds nothing to report — the brief's "Không có dữ liệu tháng này". */
export function isEmptyReport(report: FinancialReport): boolean {
  return report.incomes.length === 0 && report.expenses.length === 0;
}

export const periodStart = (period: PeriodFilter): number | null => periodBounds(period)?.[0] ?? null;
